File: javascript/takeuchi.js
const x = 12;
const y = 6;
const z = 0;
const repeatCount = 10;

function tarai(x, y, z) {
  return x <= y ?
    y :
    tarai(
      tarai(x - 1, y, z),
      tarai(y - 1, z, x),
      tarai(z - 1, x, y),
    );
}

const start = Date.now();
for (let i = 0; i < repeatCount; ++i) {
  tarai(x, y, z);
}
const end = Date.now();

console.log(end - start);

File: ruby/takeuchi.rb
x = 12
y = 6
z = 0
repeat_count = 10

def tarai(x, y, z)
  if x <= y
    y
  else
    tarai(
      tarai(x - 1, y, z),
      tarai(y - 1, z, x),
      tarai(z - 1, x, y),
    )
  end
end

start = Time.now
repeat_count.times { tarai(x, y, z) }
_end = Time.now

puts ((_end - start) * 1000).to_i
